File: frontend/src/types/dto/npa.ts
/**
 * Реестр НПА — форма ответа `GET /npa` (backend/app/schemas/npa.py).
 *
 * Срез-141: до него витрина ждала от сервера `issuer`/`status`/`effective_at`
 * и страничную обёртку, которых у ручки никогда не было, — колонки «Орган» и
 * «Статус» показывали бы прочерк на любых данных. Теперь типы списаны с
 * сервера, а «действует / утратил силу» считается по датам `valid_from`/`valid_to`.
 */

export interface NpaClauseDto {
  id: string;
  code: string;
  text: string;
}

/**
 * Срез-201: из какого ящика акт. `registry` — общий реестр платформы
 * (федеральный приказ), `own` — собственный акт этой организации. Подпись
 * словами приходит С СЕРВЕРА (`scope_title`): витрина не должна знать, что
 * "own" значит «Акт организации».
 */
export type NpaScope = "registry" | "own";

export interface NpaDto {
  id: string;
  code: string;
  title: string;
  edition: string;
  valid_from?: string | null;
  valid_to?: string | null;
  clauses: NpaClauseDto[];
  scope?: NpaScope;
  scope_title?: string;
}

export interface NpaListResponseDto {
  items: NpaDto[];
  /** Право заводить акты и редакции ОБЩЕГО реестра — у владельца платформы. */
  can_manage: boolean;
  /**
   * Срез-201: право завести СВОЙ акт. Оно шире предыдущего: свой приказ ведёт
   * любая организация. Без отдельного флага кнопка «Добавить свой акт»
   * пряталась бы ровно у тех, кому она и нужна.
   */
  can_create_own?: boolean;
}

export interface NpaClauseCreateDto {
  code: string;
  text: string;
}

export interface NpaActCreateDto {
  /** Срез-201: в какой ящик писать. По умолчанию — общий реестр. */
  scope?: NpaScope;
  code: string;
  title: string;
  edition: string;
  valid_from?: string | null;
  valid_to?: string | null;
  clauses?: NpaClauseCreateDto[];
}

/**
 * Срез-202: что изменилось между двумя редакциями акта (B.18 разд. 19.4).
 *
 * `comparable: false` — это ОТКАЗ, а не «изменений нет». У редакций, заведённых
 * до среза-202, текста нет вовсе, и показать по ним пустой список значило бы
 * сказать «закон не менялся» — после чего документы никто не пересмотрит.
 */
export type NpaClauseChangeKind = "added" | "removed" | "modified";

export interface NpaClauseChangeDto {
  code: string;
  change: NpaClauseChangeKind;
  /** Подпись словами приходит с сервера: витрина код не переводит. */
  change_title: string;
  before?: string | null;
  after?: string | null;
}

export interface NpaRevisionDiffDto {
  comparable: boolean;
  /** Причина отказа словами. Пусто, когда сравнение состоялось. */
  reason: string;
  changes: NpaClauseChangeDto[];
  summary: {
    added: number;
    removed: number;
    modified: number;
    unchanged: number;
  };
  base: { id: string; revision_code: string; title: string };
  target: { id: string; revision_code: string; title: string };
}

/**
 * Срез-203 (B.18 разд. 19.1 «owner»): кто ведёт этот акт в НАШЕЙ организации.
 *
 * Ответственный арендаторский: один и тот же приказ Минтруда ведут в разных
 * организациях разные люди. `responsible: null` — «акт никто не ведёт», и это
 * честное состояние, а не пропуск данных.
 */
export interface NpaResponsibleDto {
  user_id: string;
  name: string;
}

export interface NpaResponsibleResponseDto {
  responsible: NpaResponsibleDto | null;
  candidates: Array<{ id: string; name: string; role_label?: string }>;
}

export interface NpaRevisionCreateDto {
  revision_code: string;
  title: string;
  effective_from?: string | null;
  effective_to?: string | null;
  change_summary?: string | null;
}

export interface NpaRevisionDto extends NpaRevisionCreateDto {
  id: string;
  act_id: string;
}

export interface NpaFiltersDto {
  /** Подстрока по коду и названию; сервер списка не фильтрует — ищем на витрине. */
  search?: string;
}

/**
 * Срез-142: связи акта с сущностями арендатора — то, что читает оценка
 * влияния. До среза их заводить было нечем, и «Связанные сущности» были пусты
 * на любых данных.
 */
export type NpaBindingTarget = "document" | "template_version" | "pack";

export interface NpaBindingDto {
  id: string;
  npa_id: string;
  entity_type: NpaBindingTarget;
  entity_id: string;
  ref?: string | null;
  /**
   * Срез-197: область действия связи — кого и где касается, СЛОВАМИ
   * (`{role, site}`). Пустой объект означает «весь акт»: это отдельное
   * состояние, а не пробел.
   */
  context?: Record<string, string>;
  /** Имя сущности по-человечески: «Инструкция · ООО Ромашка», «Шаблон v3», имя пакета. */
  title: string;
  /**
   * Срез-144 (разд. 19.4): по какой редакции связь сверяли в последний раз и
   * не разошлась ли она с действующей. `stale` — «не пересмотрена».
   */
  reviewed_revision_id?: string | null;
  reviewed_revision_code?: string | null;
  stale?: boolean;
}

export interface NpaBindingCreateDto {
  entity_type: NpaBindingTarget;
  entity_id: string;
  ref?: string | null;
  /** Срез-197: кого и где касается связь. Пустой объект — «весь акт». */
  context?: { role_code?: string; site_id?: string };
}

/** Справочники области действия связи (срез-197). */
export interface NpaBindingOptionsDto {
  roles: Array<{ code: string; label: string }>;
  sites: Array<{ id: string; name: string; company_name: string }>;
}

File: backend/app/schemas/npa.py
"""Pydantic schemas for normative legal acts."""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.domains.npa.scope import REGISTRY_SCOPE, TENANT_SCOPE


class NpaClauseRead(BaseModel):
    id: str
    code: str = Field(max_length=128)
    text: str

    model_config = {
        "from_attributes": True,
    }


class NpaActRead(BaseModel):
    id: str
    code: str = Field(max_length=128)
    title: str
    edition: str
    valid_from: date | None = None
    valid_to: date | None = None
    clauses: list[NpaClauseRead] = []
    #: Срез-201: из какого ящика акт — общий реестр платформы или собственный
    #: акт арендатора. Витрина обязана их различать: «Приказ Минтруда» и «наш
    #: приказ по организации» лежат в одном списке, а весят разное.
    scope: str = REGISTRY_SCOPE
    scope_title: str = ""

    model_config = {
        "from_attributes": True,
    }


class NpaActListResponse(BaseModel):
    items: list[NpaActRead]
    #: Срез-141: может ли пришедший заводить акты и редакции. Реестр общий для
    #: всех арендаторов, поэтому право есть только у владельца платформы —
    #: витрина по этому флагу показывает или прячет кнопку «Добавить акт».
    can_manage: bool = False
    #: Срез-201: а свой собственный акт завести может? Это ДРУГОЕ право: общий
    #: реестр закрыт всем, кроме владельца платформы, а свой приказ ведёт любой
    #: арендатор. Без отдельного флага витрина прятала бы кнопку «Добавить свой
    #: акт» ровно у тех, кому она и нужна.
    can_create_own: bool = False


class NpaClauseCreate(BaseModel):
    code: str = Field(min_length=1, max_length=128)
    text: str = Field(min_length=1)


class NpaActCreate(BaseModel):
    """Срез-141: заявка на новый акт реестра — вместе с пунктами.

    Пункты идут в той же заявке, потому что отдельной ручки для них нет и не
    нужно: пункт без акта не существует, а акт без пунктов — законен (пункты
    можно не расписывать, пока на них не ссылается оценка влияния).
    """

    #: Срез-201: в какой ящик писать. Значение по умолчанию — общий реестр:
    #: так ручка ведёт себя ровно как до среза, и старые вызовы не меняют
    #: смысла молча.
    scope: Literal[REGISTRY_SCOPE, TENANT_SCOPE] = REGISTRY_SCOPE
    code: str = Field(min_length=1, max_length=128)
    title: str = Field(min_length=1, max_length=512)
    edition: str = Field(min_length=1, max_length=128)
    valid_from: date | None = None
    valid_to: date | None = None
    clauses: list[NpaClauseCreate] = []

    @model_validator(mode="after")
    def _check(self) -> NpaActCreate:
        if self.valid_from and self.valid_to and self.valid_to < self.valid_from:
            raise ValueError("valid_to не может быть раньше valid_from")
        codes = [clause.code for clause in self.clauses]
        if len(codes) != len(set(codes)):
            raise ValueError("коды пунктов внутри акта должны быть уникальны")
        return self


class NpaRevisionCreate(BaseModel):
    revision_code: str = Field(min_length=1, max_length=128)
    title: str = Field(min_length=1, max_length=512)
    effective_from: date | None = None
    effective_to: date | None = None
    change_summary: str | None = None

    @model_validator(mode="after")
    def _check(self) -> NpaRevisionCreate:
        if self.effective_from and self.effective_to and self.effective_to < self.effective_from:
            raise ValueError("effective_to не может быть раньше effective_from")
        return self


class NpaRevisionRead(BaseModel):
    id: str
    act_id: str
    revision_code: str
    title: str
    effective_from: date | None = None
    effective_to: date | None = None
    change_summary: str | None = None

    model_config = {
        "from_attributes": True,
    }


class NpaBindingContext(BaseModel):
    """Область действия связи: кого и где она касается (срез-197).

    Не свободный словарь: лишние ключи отвергаются. Свободное поле здесь уже
    было — именно из-за него сводка влияния читала то, что никто не мог
    записать, и всегда показывала ноль.
    """

    role_code: str | None = Field(default=None, max_length=64)
    site_id: str | None = Field(default=None, max_length=36)

    model_config = {"extra": "forbid"}


class NpaBindingCreate(BaseModel):
    """Срез-142: связь акта с сущностью арендатора — то, что читает оценка влияния."""

    entity_type: Literal["document", "template_version", "pack"]
    entity_id: str = Field(min_length=1, max_length=36)
    ref: str | None = Field(default=None, max_length=255)
    #: Срез-197: «кого и где касается». До него ручка ставила контекст пустым, и
    #: категории сводки, читавшие его, были вечными нулями.
    context: NpaBindingContext = Field(default_factory=NpaBindingContext)


class NpaBindingRead(BaseModel):
    id: str
    npa_id: str
    entity_type: str
    #: Область действия связи словами (срез-197): роль и площадка, если заданы.
    context: dict[str, str] = Field(default_factory=dict)
    entity_id: str
    ref: str | None = None
    #: Имя сущности для витрины — документ, шаблон или пакет по-человечески,
    #: а не идентификатор.
    title: str
    #: Срез-144 (разд. 19.4): по какой редакции связь сверяли в последний раз
    #: и не разошлась ли она с действующей. ``stale`` — «не пересмотрена».
    reviewed_revision_id: str | None = None
    reviewed_revision_code: str | None = None
    stale: bool = False
